Dockerfile

```diff
@@ -1,21 +1,86 @@
-FROM node:carbon
+# https://blog.hasura.io/an-exhaustive-guide-to-writing-dockerfiles-for-node-js-web-apps-bbee6bd2f3c4
+# https://codefresh.io/docker-tutorial/node_docker_multistage/
 
-RUN mkdir -p /parse-server
-COPY ./ /parse-server/
 
-RUN mkdir -p /parse-server/config
-VOLUME /parse-server/config
+# ------- Stage 1 - Base ---------
+FROM node:carbon-alpine as base
+# apk - https://www.cyberciti.biz/faq/10-alpine-linux-apk-command-examples/
+# RUN apk update && apk upgrade && \
+#     apk add --no-cache git bash
+# python make and g++  - add this for bcrypt 3.x 
 
-RUN mkdir -p /parse-server/cloud
-VOLUME /parse-server/cloud
+# ENV, WORKDIR & COPY commands run as USER root
+# to change ownership run chown (e.g. chown USER node /parse_server) after every command
 
+# after this refer to workdir using ./
 WORKDIR /parse-server
 
-RUN npm install && \
-    npm run build
+# specify multiple volumes in one line - reuse layers during build
+VOLUME ["parse-server/config", "parse-server/cloud"]
 
-ENV PORT=1337
+# copy all package.json-related files
+COPY package.json ./
 
-EXPOSE $PORT
 
-ENTRYPOINT ["npm", "start", "--"]
+# ------- Stage 2 - Dependencies ---------
+# base image for release stage with only prod dependencies
+FROM base AS dependencies
+# set npm configs 
+RUN npm set progress=false && npm config set depth 0
+# install production packages only
+RUN npm install --production 
+
+
+# ------- Stage 3 - Build ---------
+FROM dependencies AS build
+# install all npm required for build (and testing) - saves build time since prod dependencies are already installed
+RUN npm install 
+# copy all context into WORKDIR (/parse-server) excluding items in .dockerignore
+COPY . .
+# Need to run build explicitly as npm will not auto run scripts as ROOT
+# https://stackoverflow.com/questions/47748075/npm-postinstall-not-running-in-docker
+# https://docs.npmjs.com/misc/scripts#user
+RUN npm run prepare && npm run postinstall
+# list all dir/files  - for debugging purposes
+# RUN ls -al
+
+#  ---- UNIT TESTS AND FLOW LINT in stage 3------
```

Comment:
why keep that if it is not run? We should probably either remove or run the tests

Reply:
It seems that the tests depend on a db, so we cannot run them inside Dockerfile.
Also, the tests are already setup to run in Travis CI.
I am just leaving the comments here for future reference, in case someone wants to change this.

Reply:
So please remove it!